proposals/2023/03-15_worker-auth.md

```diff
@@ -0,0 +1,177 @@
+
+With the implementation of the token manager and build tokens, the ground work is already present to mint and validate worker auth tokens.
+
+Some of the major changes will be:
+
+- Add the ability for the `internal/token` package to mint and validate worker auth / register tokens
+- Add a `GET` registration endpoint, which will be platform admin exclusive. It will return a worker registration token.
+- Enhance the current `MustWorker` permissions check to use claims from a worker auth token with the option of also using the symmetric secret, if provided. 
+- Place the `POST` and `PUT` worker endpoints behind that improved permissions check. (note: the `PUT` endpoint will accept both registration and auth tokens in order to account for re-registration in the event the worker was offline and is still in the database. It will also accept user access token types from platform admins that want to perform an update on the worker).
+- Add a `GET` validate token endpoint. One of the drawbacks of using HMAC token signing is the fact that there is no way to check validity of any given token without it being parsed by the entity with the private key (in this case, the server). By adding a simple endpoint that validates a token came from the server the worker expected, we will be able to verify tokens from the server to the worker (e.g. canceling a build).
+
+### Worker Design
+
+The worker design is more tricky, but will look something like this:
+
+1. On start up, the worker will enter a dead loop, where it will do nothing (no pulling from queue, checking in, etc) except serve the worker API. It will then be the job of the platform administrators to "register" the worker by hitting the worker's self registration endpoint.
+2. Once the self-registration endpoint is hit, the worker will use the token gathered from the header of that incoming request and break its dead loop, thereby entering its check in loop and exec loop.
+3-a. During the check in process, the worker will exchange its registration token for an auth token, which it will then refresh on the check-in interval. The success status of the check in will be a new struct field for the worker.
+3-b. Based on the success status, the worker will proceed with its exec loop, checking before each pull whether the worker is truly verified with the server.
+4. When a build is pulled off the queue, the worker will request a build token using its auth token.
+5. In a scenario where a worker loses connection for a certain time period that renders the auth token expired, and admin is require to re-register the worker
+
+Other considerations to further improve security:
+- specification of IP address in the claims to allow restriction by IP
+- stricter expiration logic
+
+In the case where the admins would prefer to continue using the symmetric token, all the above worker design logic will use the `VELA_SERVER_SECRET` token, which will not expire and not hinder executors' ability to pull from the queue.
```

Comment:
Do you mean that the `VELA_SERVER_SECRET` can be used as a persistent worker auth token?

Reply:
Yes, the idea is that if it's provided (and there is a matching `VELA_SECRET` in the server env) then the server will honor that as a non-expiring auth method